### ISSUE TITLE
[core] Designer cannot handle classes in packages

### ISSUE THREAD

**Issue by Vampire** [pmd/pmd#1672]
When pasting a class with package statement to the Designer and pressing "Go", the following stacktrace is printed and nothing is displayed in the lower panels:

```java
Exception in thread "AWT-EventQueue-0" java.lang.NullPointerException
	at net.sourceforge.pmd.lang.java.typeresolution.ClassTypeResolver.getClassName(ClassTypeResolver.java:1448)
	at net.sourceforge.pmd.lang.java.typeresolution.ClassTypeResolver.visit(ClassTypeResolver.java:198)
	at net.sourceforge.pmd.lang.java.ast.ASTCompilationUnit.jjtAccept(ASTCompilationUnit.java:41)
	at net.sourceforge.pmd.lang.java.typeresolution.TypeResolutionFacade.initializeWith(TypeResolutionFacade.java:18)
	at net.sourceforge.pmd.lang.java.AbstractJavaHandler$5.start(AbstractJavaHandler.java:121)
	at net.sourceforge.pmd.util.designer.Designer.getCompilationUnit(Designer.java:205)
	at net.sourceforge.pmd.util.designer.Designer.getCompilationUnit(Designer.java:210)
	at net.sourceforge.pmd.util.designer.Designer.getCompilationUnit(Designer.java:198)
	at net.sourceforge.pmd.util.designer.Designer.access$300(Designer.java:127)
	at net.sourceforge.pmd.util.designer.Designer$XPathListener.actionPerformed(Designer.java:601)
	at java.desktop/javax.swing.AbstractButton.fireActionPerformed(AbstractButton.java:1967)
	at java.desktop/javax.swing.AbstractButton$Handler.actionPerformed(AbstractButton.java:2308)
	at java.desktop/javax.swing.DefaultButtonModel.fireActionPerformed(DefaultButtonModel.java:405)
	at java.desktop/javax.swing.DefaultButtonModel.setPressed(DefaultButtonModel.java:262)
	at java.desktop/javax.swing.plaf.basic.BasicButtonListener.mouseReleased(BasicButtonListener.java:279)
	at java.desktop/java.awt.Component.processMouseEvent(Component.java:6632)
	at java.desktop/javax.swing.JComponent.processMouseEvent(JComponent.java:3342)
	at java.desktop/java.awt.Component.processEvent(Component.java:6397)
	at java.desktop/java.awt.Container.processEvent(Container.java:2263)
	at java.desktop/java.awt.Component.dispatchEventImpl(Component.java:5008)
	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2321)
	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4840)
	at java.desktop/java.awt.LightweightDispatcher.retargetMouseEvent(Container.java:4918)
	at java.desktop/java.awt.LightweightDispatcher.processMouseEvent(Container.java:4547)
	at java.desktop/java.awt.LightweightDispatcher.dispatchEvent(Container.java:4488)
	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2307)
	at java.desktop/java.awt.Window.dispatchEventImpl(Window.java:2772)
	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4840)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:772)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:721)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:715)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:95)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:745)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:743)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:742)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)
```

**Comment by jsotuyod** [comment]
@Vampire what version of PMD are you using? The "Go" button has been removed almost a year ago if I recall correctly. The designer currently updates automatically as the code is changed / the XPath expression is edited.

Please, make sure you are using PMD 6.11.0.

**Comment by Vampire** [comment]
I started the `Designer` class from inside the `pmd-core` JAR.
I see now it is deprecated.
Where is the current to be found?

**Comment by jsotuyod** [comment]
Oh, yeah, that one is not even in the codebase anymore for 7.0.0 (see #1431)

There is now a dedicated pmd-ui module using JavaFX with a **much improved designer**. You will find it to be way more useful and user friendly. If running from CLI, you can use `run.sh designer` or `designer.bat` (according to your OS).

**Comment by Vampire** [comment]
Yep, found the pmd-ui module already, thx.
Now struggling with getting it to run properly.
I run it from the IDE as I'm in it anyway.

First problem: I started with Java 11 and you don't depend on OpenJFX, so on JavaFX present and it dies with missing `Application` class

Next problem: Started with Java 9, just died without any message, after a bit debugging turned out to be an NPE in the `ASTManager` because no language module was present, should maybe instead show a meaningful error message instead instead.

Next problem: Started with Java 9 and additionally the pmd-java module, again died without message, this time because of `java.lang.reflect.InaccessibleObjectException: Unable to make field private final javafx.scene.control.skin.VirtualFlow javafx.scene.control.skin.TreeViewSkin.flow accessible: module javafx.controls does not "opens javafx.scene.control.skin" to unnamed module @6dde5c8c`

Finally started with Java 1.8 and additionally the pmd-java module and now it worked.

**Comment by Vampire** [comment]
And if I paste my code in there, also nothing happens at first, until I manually select a Java version which was selected by default in the old designer.

**Comment by Vampire** [comment]
And while it is a bit unclear what the properties panel is for, if you select add or edit in the context menu, a popup without any window decorations is opened that looks a bit ugly and is not movable or abortable. :-/

**Comment by Vampire** [comment]
And when I type in an XPath expression nothing visually happens besides a very tiny `(error)` title that can easily be overlooked.
Wondering and finally finding the exception log I found

```java
net.sourceforge.pmd.util.fxdesigner.model.XPathEvaluationException: java.lang.SecurityException: class "net.sf.saxon.om.ValueRepresentation"'s signer information does not match signer information of other classes in the same package
	at net.sourceforge.pmd.util.fxdesigner.model.XPathEvaluator.evaluateQuery(XPathEvaluator.java:89)
	at net.sourceforge.pmd.util.fxdesigner.XPathPanelController.evaluateXPath(XPathPanelController.java:250)
	at net.sourceforge.pmd.util.fxdesigner.MainDesignerController.lambda$refreshXPathResults$10(MainDesignerController.java:210)
	at java.util.Optional.ifPresent(Optional.java:159)
	at net.sourceforge.pmd.util.fxdesigner.MainDesignerController.refreshXPathResults(MainDesignerController.java:210)
	at net.sourceforge.pmd.util.fxdesigner.XPathPanelController.lambda$beforeParentInit$4(XPathPanelController.java:137)
	at org.reactfx.util.NonAccumulativeStreamNotifications.lambda$head$0(NotificationAccumulator.java:134)
	at org.reactfx.ObservableBase.notifyObservers(ObservableBase.java:68)
	at org.reactfx.ObservableBase.notifyObservers(ObservableBase.java:57)
	at org.reactfx.ProperEventStream.emit(ProperEventStream.java:18)
	at org.reactfx.EventStream$1.lambda$observeInputs$1(EventStream.java:424)
	at org.reactfx.util.NonAccumulativeStreamNotifications.lambda$head$0(NotificationAccumulator.java:134)
	at org.reactfx.ObservableBase.notifyObservers(ObservableBase.java:68)
	at org.reactfx.ObservableBase.notifyObservers(ObservableBase.java:57)
	at org.reactfx.ProperEventStream.emit(ProperEventStream.java:18)
	at org.reactfx.EventStreams$5.lambda$observeInputs$0(EventStreams.java:144)
	at org.reactfx.value.ChangeListenerWrapper.accept(Val.java:786)
	at org.reactfx.util.AbstractReducingStreamNotifications.lambda$head$0(NotificationAccumulator.java:248)
	at org.reactfx.ObservableBase.notifyObservers(ObservableBase.java:68)
	at org.reactfx.ObservableBase.notifyObservers(ObservableBase.java:57)
	at org.reactfx.value.ValBase.invalidate(ValBase.java:32)
	at org.reactfx.value.MappedVal.lambda$connect$0(MappedVal.java:28)
	at org.reactfx.value.InvalidationListenerWrapper.accept(Val.java:765)
	at org.reactfx.util.AbstractReducingStreamNotifications.lambda$head$0(NotificationAccumulator.java:248)
	at org.reactfx.ObservableBase.notifyObservers(ObservableBase.java:68)
	at org.reactfx.ObservableBase.notifyObservers(ObservableBase.java:57)
	at org.reactfx.value.ValBase.invalidate(ValBase.java:32)
	at org.reactfx.value.ValWrapper.lambda$connect$0(ValWrapper.java:21)
	at com.sun.javafx.binding.ExpressionHelper$SingleInvalidation.fireValueChangedEvent(ExpressionHelper.java:137)
	at com.sun.javafx.binding.ExpressionHelper.fireValueChangedEvent(ExpressionHelper.java:81)
	at javafx.beans.property.ReadOnlyObjectPropertyBase.fireValueChangedEvent(ReadOnlyObjectPropertyBase.java:74)
	at javafx.beans.property.ReadOnlyObjectWrapper.fireValueChangedEvent(ReadOnlyObjectWrapper.java:102)
	at javafx.beans.property.ObjectPropertyBase.markInvalid(ObjectPropertyBase.java:112)
	at javafx.beans.property.ObjectPropertyBase.set(ObjectPropertyBase.java:146)
	at javafx.scene.control.ToggleGroup$3.set(ToggleGroup.java:137)
	at javafx.scene.control.ToggleGroup$3.set(ToggleGroup.java:119)
	at javafx.scene.control.ToggleGroup.selectToggle(ToggleGroup.java:149)
	at javafx.scene.control.RadioMenuItem$2.invalidated(RadioMenuItem.java:193)
	at javafx.beans.property.BooleanPropertyBase.markInvalid(BooleanPropertyBase.java:109)
	at javafx.beans.property.BooleanPropertyBase.set(BooleanPropertyBase.java:144)
	at javafx.scene.control.RadioMenuItem.setSelected(RadioMenuItem.java:180)
	at com.sun.javafx.scene.control.skin.ContextMenuContent$MenuItemContainer.doSelect(ContextMenuContent.java:1401)
	at com.sun.javafx.scene.control.skin.ContextMenuContent$MenuItemContainer.lambda$createChildren$343(ContextMenuContent.java:1358)
	at com.sun.javafx.event.CompositeEventHandler$NormalEventHandlerRecord.handleBubblingEvent(CompositeEventHandler.java:218)
	at com.sun.javafx.event.CompositeEventHandler.dispatchBubblingEvent(CompositeEventHandler.java:80)
	at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:238)
	at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:191)
	at com.sun.javafx.event.CompositeEventDispatcher.dispatchBubblingEvent(CompositeEventDispatcher.java:59)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:58)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.EventUtil.fireEventImpl(EventUtil.java:74)
	at com.sun.javafx.event.EventUtil.fireEvent(EventUtil.java:54)
	at javafx.event.Event.fireEvent(Event.java:198)
	at javafx.scene.Scene$MouseHandler.process(Scene.java:3757)
	at javafx.scene.Scene$MouseHandler.access$1500(Scene.java:3485)
	at javafx.scene.Scene.impl_processMouseEvent(Scene.java:1762)
	at javafx.scene.Scene$ScenePeerListener.mouseEvent(Scene.java:2494)
	at com.sun.javafx.tk.quantum.GlassViewEventHandler$MouseEventNotification.run(GlassViewEventHandler.java:381)
	at com.sun.javafx.tk.quantum.GlassViewEventHandler$MouseEventNotification.run(GlassViewEventHandler.java:295)
	at java.security.AccessController.doPrivileged(Native Method)
	at com.sun.javafx.tk.quantum.GlassViewEventHandler.lambda$handleMouseEvent$354(GlassViewEventHandler.java:417)
	at com.sun.javafx.tk.quantum.QuantumToolkit.runWithoutRenderLock(QuantumToolkit.java:389)
	at com.sun.javafx.tk.quantum.GlassViewEventHandler.handleMouseEvent(GlassViewEventHandler.java:416)
	at com.sun.glass.ui.View.handleMouseEvent(View.java:555)
	at com.sun.glass.ui.View.notifyMouse(View.java:937)
	at com.sun.glass.ui.win.WinApplication._runLoop(Native Method)
	at com.sun.glass.ui.win.WinApplication.lambda$null$148(WinApplication.java:191)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.SecurityException: class "net.sf.saxon.om.ValueRepresentation"'s signer information does not match signer information of other classes in the same package
	at java.lang.ClassLoader.checkCerts(ClassLoader.java:898)
	at java.lang.ClassLoader.preDefineClass(ClassLoader.java:668)
	at java.lang.ClassLoader.defineClass(ClassLoader.java:761)
	at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142)
	at java.net.URLClassLoader.defineClass(URLClassLoader.java:467)
	at java.net.URLClassLoader.access$100(URLClassLoader.java:73)
	at java.net.URLClassLoader$1.run(URLClassLoader.java:368)
	at java.net.URLClassLoader$1.run(URLClassLoader.java:362)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.net.URLClassLoader.findClass(URLClassLoader.java:361)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:335)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	at net.sourceforge.pmd.lang.rule.XPathRule.initRuleQueryBasedOnVersion(XPathRule.java:145)
	at net.sourceforge.pmd.lang.rule.XPathRule.initXPathRuleQuery(XPathRule.java:128)
	at net.sourceforge.pmd.lang.rule.XPathRule.getRuleChainVisits(XPathRule.java:151)
	at net.sourceforge.pmd.lang.rule.AbstractRule.isRuleChain(AbstractRule.java:310)
	at net.sourceforge.pmd.lang.rule.AbstractRuleChainVisitor.initialize(AbstractRuleChainVisitor.java:158)
	at net.sourceforge.pmd.lang.rule.AbstractRuleChainVisitor.visitAll(AbstractRuleChainVisitor.java:63)
	at net.sourceforge.pmd.RuleChain.apply(RuleChain.java:67)
	at net.sourceforge.pmd.RuleSets.apply(RuleSets.java:140)
	at net.sourceforge.pmd.util.fxdesigner.model.XPathEvaluator.evaluateQuery(XPathEvaluator.java:84)
	... 73 more
````

Switching the setting to XPath 1.0 works, with the default 2.0 always this exception comes.

**Comment by Vampire** [comment]
And typing a `[` which here (German keyboard) means pressing AltGr+8 activates the menu bar and you cannot type on. :-/

**Comment by oowekyala** [comment]
Thanks for your comments, that's very valuable to improve user experience on the designer. We've not put much effort into smoothing IDE setup, the [launch script](https://github.com/pmd/pmd/blob/master/pmd-dist/src/main/resources/scripts/run.sh) included in the binary distribution being the recommended way to start the app.

Running with Java 9 and above requires some additionnal VM parameters, which the launch script adds itself: 
https://github.com/pmd/pmd/blob/c4ce925997917e741c9d53258c0090b1f8c327d8/pmd-dist/src/main/resources/scripts/run.sh#L116-L118
We can definitely do better on the documentation side.  

 Running with no language modules is not a scenario I've even thought about but we should definitely handle that in a user-friendly way

> And while it is a bit unclear what the properties panel is for, if you select add or edit in the context menu, a popup without any window decorations is opened that looks a bit ugly and is not movable or abortable. :-/

There's already [a ticket](https://github.com/pmd/pmd/projects/2#card-16580557) on the designer project about that. The solution I'd like to go for is blocked by #1630 which has been stalling for a while. But it will be improved sooner or later. The lack of decorations was originally intended as a way to prevent user from closing while the property has invalid metadata, but it's definitely annoying.

> Caused by: java.lang.SecurityException: class "net.sf.saxon.om.ValueRepresentation"'s signer information does not match signer information of other classes in the same package

I have no idea what might have caused that. Never encountered that. I'll try to dig on it. What's your IDE setup exactly? How did you import the project? 

> And typing a `[` which here (German keyboard) means pressing AltGr+8 activates the menu bar and you cannot type on. :-/
> And if I paste my code in there, also nothing happens at first, until I manually select a Java version which was selected by default in the old designer.

Those will be fixed shortly

**Comment by Vampire** [comment]
A missing variable is also like the xpath syntax, you paste in some XPath from the doc to see the instance in your code it matched, and you just get the unobtrusive `(error)` and have to find the exceptions log to see the error while the result panel could show the message that the variable is missing.

Besides that, changing the properties (yes, found out now what it is for :-D) does not trigger a re-evaluate, you have to change the code or xpath expression to trigger a re-evaluate.

**Comment by Vampire** [comment]
And is it expected that functions like `typeIsExactly` and alike are not available in the designer?

**Comment by oowekyala** [comment]
>  For example if there is a syntactical error in the XPath expression, the message could be displayed in the result window instead of an unobtrusive black `(error)`.

That's a good idea. I'll try to work out a prototype

> And is it expected that functions like `typeIsExactly` and alike are not available in the designer?

What do you mean? You can use them in XPath queries if that's your question. The type info was available in the XPath attributes tab until #1558, it's reintroduced by #1630. In all cases for it to work properly you have to fill-in your user classpath in the `Configuration` menu.


I'll try to break up all your comments into tickets for the designer project board

**Comment by oowekyala** [comment]
> You can always have an OK button for applying the changes and treat the close button as abort and to make sure no work is lost ask for confirmation if something was changed in the dialog.
Or you could refuse to close the popup even when the close button was pressed if the content is not valid if you don't want to have the OK / Cancel semantic.

I'd like to try getting rid of the popup entirely, possibly rather using a [PopOver](http://fxexperience.com/controlsfx/features/#popover) from ControlsFX. I've not yet experimented with it. If that's not satisfactory a popup with Ok/Cancel semantics seems the best choice I think.

**Comment by Vampire** [comment]
>> And is it expected that functions like `typeIsExactly` and alike are not available in the designer?

> What do you mean? You can use them in XPath queries if that's your question. The type info was available in the XPath attributes tab until #1558, it's reintroduced by #1630. In all cases for it to work properly you have to fill-in your user classpath in the `Configuration` menu.

If I try to use them, I only get another exception in the log that the function does not exist, because `net.sourceforge.pmd.lang.ast.xpath.DefaultASTXPathHandler#initialize()` is never called and thus those custom functions are never registered. Or is this again an issue because I just start from IDE?

> I'll try to break up all your comments into tickets for the designer project board

I'm sorry, I usually tend to make too much tickets, not too less, but as some things were more questions here, I just didn't think about it and wrote on, usually I make one issue per topic.

> I'd like to try getting rid of the popup entirely, possibly rather using a [PopOver](http://fxexperience.com/controlsfx/features/#popover) from ControlsFX. I've not yet experimented with it. If that's not satisfactory a popup with Ok/Cancel semantics seems the best choice I think.

Probably also ok, or maybe just make the table have editable cells. Except you want to hide more information or maybe want to make the type column hidden by default which would maybe also make sense.

**Comment by oowekyala** [comment]
> If I try to use them, I only get another exception in the log that the function does not exist, because `net.sourceforge.pmd.lang.ast.xpath.DefaultASTXPathHandler#initialize()` is never called and thus those custom functions are never registered. Or is this again an issue because I just start from IDE?

This deserves its own issue I think. Mind creating it with the full log output of the exception?

> Probably also ok, or maybe just make the table have editable cells. Except you want to hide more information or maybe want to make the type column hidden by default which would maybe also make sense.

The validation framework we use (from ControlsFX) doesn't handle table views so we couldn't have visual validation on editable cells without implementing it ourselves. This can be done, but at the time I introduced the property table view I wanted to avoid the trouble. Getting it to be visually consistent with the vanilla ControlsFX validation would be quite hard I think. But that's not to say it's off the table

**Comment by Vampire** [comment]
> This deserves its own issue I think. Mind creating it with the full log output of the exception?

Not at all: #1681 

> The validation framework we use (from ControlsFX) doesn't handle table views so we couldn't have visual validation on editable cells without implementing it ourselves.

Ah, I see.

**Comment by marquiswang** [comment]
FYI, I ran into the 
```
Caused by: java.lang.SecurityException: class "net.sf.saxon.om.ValueRepresentation"'s signer information does not match signer information of other classes in the same package
```

This turned out to be a class path collision. I have checkstyle 10.6.0 and spotbugs 4.7.3 on the class path as well, and those depend on net.sf.saxon:Saxon-HE:jar:11.4, which has the same packages/classes.
